apps/gimbtonomy/main.cpp

```diff
@@ -47,33 +76,82 @@ void handleSetNeoPixelColor(CANMsg *p_newMsg) {
   }
 }
 
-void processCANMsg(CANMsg *p_newMsg) {
-  switch (p_newMsg->id) {
-    case CANID::SET_NEOPIXEL:
-      // Send an acknowledgement CANMsg back to the Jetson
-      printf("Sending acknowledgement message\r\n");
-      txMsg.clear();
-      txMsg.id = CANID::NEOPIXEL_ACK;
-      txMsg << true;
-      can.write(txMsg);
-      printf("Updating neo pixels\r\n");
-      handleSetNeoPixelColor(p_newMsg);
-      break;
-    default:
-      printf("Recieved unimplemented command for neopixels\r\n");
-      break;
+// CAN Message Processor
+void rxCANProcessor() {
+  // Position
+  float pan_pos = 0, pan_speed = 0, pitch_pos = 0;
+
+  while (true) {
+    if (can1.read(rxMsg)) {
+      switch (rxMsg.id) {
+        case CANID::SET_GIMBAL_PAN_POS:
+          rxMsg.getPayload(pan_pos);
+          can_debug.write(1);
+          break;
+
+        case CANID::SET_GIMBAL_PAN_SPEED:
+          rxMsg.getPayload(pan_speed);
+          break;
+
+        case CANID::SET_GIMBAL_PITCH_POS:
+          rxMsg.getPayload(pitch_pos);
+          break;
+
+        case CANID::SET_NEOPIXEL:
+          // Send an acknowledgement CANMsg back to the Jetson
+          printf("Sending neopixel acknowledgement message\r\n");
+          txMsg.clear();
+          txMsg.id = CANID::NEOPIXEL_ACK;
+          txMsg << true;
+          can1.write(txMsg);
+          printf("Updating neo pixels\r\n");
+          handleSetNeoPixelColor(&rxMsg);
+          break;
+
+        default:
+          break;
+      }
+    }
+
+    pos_data.put(&pan_pos);
+    pos_data.put(&pan_speed);
+    pos_data.put(&pitch_pos);
+
+    event_flags.set(MOVE_FLAG);  // Set event flag for moveThread to run
+
+    ThisThread::sleep_for(5);  // Sleep CAN processor for 5 ms
   }
 }
 
-// main() runs in its own thread in the OS
-int main() {
-  printf("Beginning neopixel fw app.\r\n");
-  initCAN();
-  while (1) {
-    if (can.read(rxMsg)) {
-      processCANMsg(&rxMsg);
-      rxMsg.clear();
-    }
+// Move servos to determined locations
+
+void move() {
+  while (true) {
+    event_flags.wait_any(MOVE_FLAG);  // moveThread does not sleep, but only executes if signaled by can processor
+
+    // Set pitch pos
+    pitchServo.setPosition(*((float *)(pos_data.get().value.p)));  // Cast float to evt ptr and dereference float ptr
```

Comment:
Careful with this kind of design, where you are depending on the order of queue

Reply:
Yeah I'll change that, also don't think we need a separate thread for moving the gimbal